File: Solutions/LeetCode/3514-NumberOfUniqueXorTripletsII/Solution.cs
using System;

public class Solution
{
    public int UniqueXorTriplets(int[] nums)
    {
        int limit = 2048;
        
        // s1 tracks all pair XORs (nums[i] ^ nums[j]) for i <= j
        bool[] s1 = new bool[limit];
        int n = nums.Length;
        for (int i = 0; i < n; i++)
        {
            for (int j = i; j < n; j++)
            {
                s1[nums[i] ^ nums[j]] = true;
            }
        }
        
        // s2 tracks all triplet XORs (pair_xor ^ nums[k])
        bool[] s2 = new bool[limit];
        for (int x = 0; x < limit; x++)
        {
            if (s1[x])
            {
                foreach (int num in nums)
                {
                    s2[x ^ num] = true;
                }
            }
        }
        
        int count = 0;
        for (int i = 0; i < limit; i++)
        {
            if (s2[i]) count++;
        }
        
        return count;
    }

    static void Main()
    {
        var sol = new Solution();
        int result = sol.UniqueXorTriplets(new int[] { 1, 3, 5 });
        Console.WriteLine("Example [1, 3, 5] Unique Triplet XOR Count: " + result);
    }
}

File: Solutions/LeetCode/3514-NumberOfUniqueXorTripletsII/Solution.py
from typing import List

class Solution:
    def UniqueXorTriplets(self, nums: List[int]) -> int:
        # Since nums[i] <= 1500, the maximum possible XOR value is < 2048
        limit = 2048
        
        # s1 tracks all pair XORs (nums[i] ^ nums[j]) for i <= j
        s1 = [False] * limit
        n = len(nums)
        for i in range(n):
            for j in range(i, n):
                s1[nums[i] ^ nums[j]] = True
                
        # s2 tracks all triplet XORs (pair_xor ^ nums[k])
        s2 = [False] * limit
        for x in range(limit):
            if s1[x]:
                for num in nums:
                    s2[x ^ num] = True
                    
        return sum(s2)

if __name__ == "__main__":
    sol = Solution()
    print("Example [1, 3, 5] Unique Triplet XOR Count:", sol.UniqueXorTriplets([1, 3, 5]))
